### DISCUSSION TITLE
Turning line numbering in docstrings -- is it possible?

### DISCUSSION THREAD

**Discussion by pykhaz** [squidfunk/mkdocs-material#7683] [Q&A] [answered]
I have noticed line numbers added in the generated html document, where, ideally, they shouldn't have been. In the docstring of __init__.py file there is:
```
# (...)
Source files:

    └── src
        └── khcolors
            ├── __about__.py
            ├── colors_util.py
            ├── __init__.py
            ├── lib.py
            └── __main__.py
# (...)
```

This piece of text is rendered

![tree_linenumbers_unnecessary](https://github.com/user-attachments/assets/09a7a6de-33a4-46f5-9229-741251af6e3b)

-- below regular text "Source files:", there is a nice block of code, but each and every line of it is preceded with a number, e.g.

`1.    └── src`

Is it possible to turn line numbering off in this, and similar, cases?

**Comment by kamilkrzyskow** [comment, marked as answer]
Hello @pykhaz,
based on the [`linenums` option](https://facelessuser.github.io/pymdown-extensions/extensions/highlight/#options) description I read it as either:

> Enable line numbers globally for block code. This will be ignored for inline code. If set to False line numbers will be disabled globally and can not be turned on, not even per code block.
- enabled globally, without the option to turn it off for specific code blocks
- "not set" globally, **with** the option to turn it on for specific code blocks
- disabled globally, without the option to turn it on for specific code blocks
🤔 

However, the first bullet point is not explicitly stated, so I checked the source code out:
- https://github.com/facelessuser/pymdown-extensions/blob/20be2fc43981a376923482d7be447a2e7c910cc3/pymdownx/highlight.py#L336-L339

```py
linenums_enabled = (
    (self.linenums and linestart != 0) or
    (self.linenums is not False and linestart > 0)
) and not inline > 0
```

So I think you should be able to disable the linenums per code block with the `linestart` attribute set to 0 🤔 

````
```txt linestart=0
└── src
    └── khcolors
        ├── __about__.py
        ├── colors_util.py
        ├── __init__.py
        ├── lib.py
        └── __main__.py
```
````

**Reply by pykhaz** [reply]
It worked like a ***charm***! Great thanks :smiley:

**Comment by alexvoss** [comment]
@kamilkrzyskow has beaten me to it but the option should be `linenums`, not `linestart`:

```
    ```python linenums="0"
    print("Hello World!!!")
    ```

        print("Hello World!!!")
```

<img width="781" alt="image" src="https://github.com/user-attachments/assets/28d62717-0940-4ac7-bfbb-447fc24dfc02">

**Reply by kamilkrzyskow** [reply]
Oh okay, I assumed the `highlight` function gets the settings from the code block header, but it actually doesn't.
It turned out, that for each code block the whole `Highlight` class is created, so now the option description is more confusing, as a global False doesn't disable the option to enable it locally for 1 code block 🤔:
- https://github.com/facelessuser/pymdown-extensions/blob/20be2fc43981a376923482d7be447a2e7c910cc3/pymdownx/highlight.py#L510-L535

Sorry for butting in when you worked on the answer. I think, I set the "Answer in progress" badge, but I don't remember right now, and I totally missed you editing the OP 😅

**Reply by alexvoss** [reply]
No worries at all. The "Answer in progress" thing was useful so rarely that I did not use it soon after we invented it. No harm done, I learned something ;o)
